### PR TITLE
drivers: atmel: update rstc timer pio and watchdog drivers for Microchip sama7g54

### PR DESCRIPTION
**Pull request by TonyHan11**

### CODE REVIEW

File: core/drivers/atmel_shdwc.c [hunk 1]
```diff
@@ -178,8 +196,27 @@ static TEE_Result atmel_shdwc_probe(const void *fdt, int node,
...
+
 static const struct dt_device_match atmel_shdwc_match_table[] = {
-	{ .compatible = "atmel,sama5d2-shdwc" },
+	{
+		.compatible = "atmel,sama5d2-shdwc",
+		.compat_data = &sama5d2_compat
+	},
+	{
+		.compatible = "microchip,sama7g5-shdwc",
+		.compat_data = &sama7g54_compat,
```
Comment (jforissier) [inline core/drivers/atmel_shdwc.c:218]: Minor: would `sama7g5_compat` be a better name? I mean, the "compatible" is "g5" not "g54".

File: core/drivers/atmel_shdwc.c [hunk 2]
```diff
@@ -34,6 +34,19 @@
 #define SLOW_CLK_FREQ		32768ULL
 #define DBC_PERIOD_US(x)	DIV_ROUND_UP((1000000ULL * (x)), SLOW_CLK_FREQ)
 
+struct ddrc_reg_config {
+	uint32_t type_offset;
+	uint32_t type_mask;
+};
+
+/*
+ * @shdwc_as    Peripheral SHDWC Always Secure ()
```
Comment (jforissier) [inline core/drivers/atmel_shdwc.c:43]: Drop the parentheses or add some explanation

File: core/drivers/atmel_shdwc.c [hunk 3]
```diff
@@ -34,6 +34,19 @@
...
+ */
+struct shdwc_compat {
+	uint8_t shdwc_as;
```
Comment (jforissier) [inline core/drivers/atmel_shdwc.c:46]: What is the meaning of this field? Why `uint8_t` and not `bool`? Please add a comment.

File: core/drivers/atmel_shdwc.c [hunk 4]
```diff
@@ -34,6 +34,19 @@
 #define SLOW_CLK_FREQ		32768ULL
 #define DBC_PERIOD_US(x)	DIV_ROUND_UP((1000000ULL * (x)), SLOW_CLK_FREQ)
 
+struct ddrc_reg_config {
+	uint32_t type_offset;
+	uint32_t type_mask;
+};
```
Comment (jforissier) [inline core/drivers/atmel_shdwc.c:40]: This needs documenting.

File: core/drivers/atmel_shdwc.c [hunk 5]
```diff
@@ -156,11 +170,15 @@ static TEE_Result atmel_shdwc_probe(const void *fdt, int node,
 	if (fdt_get_status(fdt, node) != DT_STATUS_OK_SEC)
 		return TEE_ERROR_BAD_PARAMETERS;
 
-	matrix_configure_periph_secure(AT91C_ID_SYS);
+	if (compat->shdwc_as == 0)
```
Comment (jforissier) [inline core/drivers/atmel_shdwc.c:173]: `if (!compat->shdwc_as) {`

File: core/drivers/atmel_shdwc.c [hunk 6]
```diff
@@ -178,8 +196,27 @@ static TEE_Result atmel_shdwc_probe(const void *fdt, int node,
 	return sama5d2_pm_init(fdt, shdwc_base);
 }
 
+static const struct shdwc_compat sama5d2_compat = {
+	.shdwc_as = 0,
+	.ddrc = {
+		.type_offset = AT91_DDRSDRC_MDR,
+		.type_mask = AT91_DDRSDRC_MD,
```
Comment (jforissier) [inline core/drivers/atmel_shdwc.c:203]: Only `type_offset` is used, it is a bit weird, the code looks incomplete. I suppose these values are intended to be used in the `io_read32()` call in the probe function? Then please update accordingly.

**Review by jforissier — COMMENTED** [review]
Comments for "drivers: atmel_shdwc: update to compatible with sama7g5".

**Review by jforissier — COMMENTED** [review]
For "drivers: atmel_pio: update to compatible with sama7g5":

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

File: core/drivers/atmel_tcb.c [hunk 7]
```diff
@@ -50,7 +50,12 @@
 #define TCB_WPMR		0xe4
 #define  TCB_WPMR_WAKEY		0x54494d
 
+#ifdef CFG_DRIVERS_SAMA7G5_CLK
```
Comment (jforissier) [inline core/drivers/atmel_tcb.c:53]: Can't this be detected at runtime (from the DT perhaps?). Then you could avoid this new CFG.

**Review by jforissier — COMMENTED** [review]
Comment on "drivers: atmel_tcb: update to compatible with sama7g5".

**Review by jforissier — COMMENTED** [review]
For "drivers: atmel_rstc: add the function to control sama7g5's USB reset":

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

File: core/drivers/atmel_tcb.c (hunk 7 above)
Comment (TonyHan11) [reply core/drivers/atmel_tcb.c]: Thank you. `CFG_DRIVERS_SAMA7G5_CLK` has been updated to `CFG_SAMA7G5`.
Fixed others according to the comments and tags added to commit atmel_pio & atmel_rstc.

Comment (jforissier) [reply core/drivers/atmel_tcb.c]: Good enough for now, thanks.

**Comment by jforissier** [comment]
`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

File: core/drivers/atmel_tcb.c [hunk 8]
```diff
@@ -50,7 +50,12 @@
 #define TCB_WPMR		0xe4
 #define  TCB_WPMR_WAKEY		0x54494d
 
+#ifdef CFG_SAMA7G5
+static const char * const tcb_clocks[] = {"t0_clk", "t1_clk", "t2_clk",
+					  "slow_clk"};
```
Comment (etienne-lms) [inline core/drivers/atmel_tcb.c:55]: nitpicking: could you add space chars after `{` and before `}`.
suggestion:
```c
static const char * const tcb_clocks[] = {
	"t0_clk", "t1_clk", "t2_clk", "slow_clk"
};

**Review by etienne-lms — COMMENTED** [review]
`Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>` with a minor comment.

File: core/drivers/atmel_tcb.c (hunk 8 above)
Comment (TonyHan11) [reply core/drivers/atmel_tcb.c]: Updated and added the review tags. Thanks.

**Comment by TonyHan11** [comment]
@jforissier review tags added, thank you.

**Pull request merged**